proposals/0333-with-memory-rebound.md

```diff
@@ -128,15 +114,43 @@ var points = Array<CGPoint>(unsafeUninitializedCapacity: data.count/MemoryLayout
 
 ## Proposed solution
 
-We propose to lift the restriction that the strides of `T` and `Pointee` must be equal.
-This means that it will now be considered correct to re-bind from a homogeneous aggregate type to the type of its constitutive elements,
-as they are layout compatible, even though their stride is different.
+`withMemoryRebound` is currently defined for `UnsafePointer`, `UnsafeMutablePointer`,
+`UnsafeBufferPointer` and `UnsafeMutableBufferPointer`.
+The type to which the memory is bound by the `Pointer` types is called `Pointee`,
+while it is `Element` for the `BufferPointer` types.
+For simplicity the following discussion calls both `Pointee`.
+
+In the general case, the runtime performs housekeeping tasks when initializing, deinitializing or updating a value of a type.
+Initializing and deinitialization of a type that is or contains a reference type means that type-specific code is executed,
+and therefore in general data cannot be accessed as another type.
+
+`withMemoryRebound` can be used safely with pairs of types `Pointee` and `T` that do _not_ require initialization or deinitialization.
+These types do not yet have a formal name in Swift,
+but are referred to as "trivial" types in some API documentation.
+
+In order to safely use `withMemoryRebound`, the current rule
+is that the destination type, `T`, must be _layout equivalent_ with `Pointee`.
+To this we add that, as an alternative, `T` can be a homogeneous aggregate of `Pointee`, or `Pointee` can be a homogeneous aggregate of `T`.
+
+Two types A and B are layout equivalent when they are, for example:
+- identical types;
+- one is a typealias for the other;
+- trivial scalar types with the same size and alignment, including floating-point and integer types;
+- one is a class type, and the other is an `AnyObject` existential;
+- pointer types, such as `UnsafePointer` and `OpaquePointer`;
+- one is a frozen struct with a single stored property, the other is the type of its stored property;
+- one is a frozen enum with a single case, the other is the associated value of the enum's single case;
```

Comment:
I hope people don't think "enum with a single case" means "enum with a single payload case". e.g. `Optional<T>` is not layout equivalent with `T` (unless `T` is a reference). I'm not really sure where single-case enums come up to be honest. So maybe we should just remove  this rule.

Reply:
I've made single-case enums for specific `Error` implementations, but that's not very interesting.

Reply:
Removing that example sounds fine to me.